Ensemble-Model.md

```diff
@@ -0,0 +1,1066 @@
+    (2, 0, 1, 0) 0.651445337683
+    (2, 0, 1, 1) 0.684947350383
+    (2, 0, 1, 2) 0.705463845082
+    (2, 0, 2, 0) 0.650420413684
+    (2, 0, 2, 1) 0.680590889828
+    (2, 0, 2, 2) 0.704830022379
+    (2, 1, 0, 0) 0.653184699287
+    (2, 1, 0, 1) 0.690123263589
+    (2, 1, 0, 2) 0.706156340238
+    (2, 1, 1, 0) 0.651128308542
+    (2, 1, 1, 1) 0.68474123067
+    (2, 1, 1, 2) 0.705355728315
+    (2, 1, 2, 0) 0.650086878921
+    (2, 1, 2, 1) 0.68035786005
+    (2, 1, 2, 2) 0.704696794839
+    (2, 2, 0, 0) 0.652823591273
+    (2, 2, 0, 1) 0.689644935857
+    (2, 2, 0, 2) 0.705832584758
+    (2, 2, 1, 0) 0.650846311386
+    (2, 2, 1, 1) 0.684340948629
+    (2, 2, 1, 2) 0.705035266025
+    (2, 2, 2, 0) 0.649844424588
+    (2, 2, 2, 1) 0.680023236015
+    (2, 2, 2, 2) 0.704397272961
+
+
+## Comparison to old the method
+
+To validate if the MAP numbers that we get are any good, it's useful to compare them to the results of the previous recommendation process. The following is a minimal reimplementation of this `RecommendationManager`. It's used here because we want to use our masked data instead of the data fetched from HBase.
+
+
+```python
+class RecommendationManager:
+    def __init__(self, recommenders):
+        self.recommenders = recommenders
+        
+    def recommend(self, client_data, limit):
+        recommendations = []
+        
+        for r in self.recommenders:
+            recommendations += r.recommend(client_data, limit)
+            
+            if len(recommendations) >= limit:
+                break
+            
+        return recommendations[:limit]
+```
+
+This helper function is similar to `map(superlist.index, sublist`) but ignores elements from the sublist that don't appear in the superlist.
+
+
+```python
+def list_elements_to_indices(superlist, sublist):
+    result = []
+    
+    for a in sublist:
+        for i, b in enumerate(superlist):
+            if a == b:
+                result.append(i)
+                break
+        
+    return result
+```
+
+
+```python
+def evaluate_recommendation_manager(mngr):
+    return 1 - training_masked\
+        .map(lambda user: (user, mngr.recommend(user, 10)))\
+        .map(lambda (user, recommendations): average_precision(user, list_elements_to_indices(whitelist, recommendations)))\
+        .mean()
+```
+
+As we can see, the previous recommendation manager performs much worse than our new model:
+
+
+```python
+mngr = RecommendationManager([recommenders["legacy"], recommenders["collaborative"], recommenders["similarity"], recommenders["locale"]])
+evaluate_recommendation_manager(mngr)
+```
+
+
+
+
+    0.94540781980452748
+
+
+
+However, this comparison is a little bit unfair. The locale recommender is generally extremely useful and can be used as a better baseline. With this ordering (where nearly only the locale recommender is queried), we get a much more comparable result. The results are now in the same ballpark and the ensemble is better by around 2%.
```

Comment:
This is also a bit circular logic as the locale recommender uses segregates counts of addons installed per locale predict the addons to recommend per locale :)

The same data source is used to generate the counts that locale_recommender uses as is used here as an evaluation set. Therefore the generalizability only holds so long as locale-specific trends are static. Also to be noted is that the locale recommender is biased towards the most popular addons by virtue of its construction so the performance over `useful_clients` with a larger number of installed addons. This may be especially relevant when new users arrive with popular addons already installed but expect interesting recommendations.